### ISSUE TITLE
Support y4m input in cjxl

### ISSUE THREAD

**Issue by y-guyon** [libjxl/libjxl#2663]
Support for y4m file format in `cjxl` was added in #25 and removed in #614.  
If I understand correctly, #837 added the missing already downsampled input option to the libjxl API so maybe #25 could be brought back?

**Comment by jonsneyers** [comment]
No, that option is still for RGB input that is completely downsampled, not for YCbCr input.

What would be the use case for y4m input? While jxl supports YCbCr for jpeg recompression, it isn't using that colorspace when doing normal lossy compression. So there isn't really an advantage to giving input in YCbCr, it would have to be converted to RGB first anyway. Except if you want to do lossless — but then I guess you'd also need a way to decode without undoing the YCbCr?

Before we add anything, it would be good to know what exactly is needed...

**Comment by y-guyon** [comment]
Thanks for your fast answer.

> Except if you want to do lossless

Yes. We would like to compare the lossless coding performance of multiple codecs on natural images and the current data set is stored as `C420jpeg` y4m files.  
The considered alternative without changing the source data set format is to go through an intermediate lossless conversion before inputting it to cjxl, such as EXR or PNG 16-bit. Assuming ffmpeg converts from y4m to EXR or PNG properly (meaning in a lossless roundtrip way), our concern would still be that libjxl would not have the full context to best compress the data (which was YUV 4:2:0 limited range to begin with).

> but then I guess you'd also need a way to decode without undoing the YCbCr?

If it is guaranteed to be lossless, there is no need to decode it, except for verification and decompression timing measures. But y4m export would be convenient for sure.

**Comment by jonsneyers** [comment]
There is no way (I think) to convert losslessly from yuv420 to rgb, except maybe if you bump up the bit depth.
But if the data is yuv420 to begin with, it will be bad for compression to not keep it as it is.

JPEG XL can only represent full range YCbCr (the way it is in JPEG), not limited/tv-range.

Making an API for YCbCr input/output is not so trivial. There are several options for the data layout: planar (first all Y, then all Cb, then all Cr) or interleaved (e.g. YYYYCbCr in the case of 4:2:0). In the current API we're assuming the 3 color components have the same dimensions, but that's not the case when there's chroma subsampling, so we'll have to come up with some alternative way to pass buffers then. And on the decode side, we could add an option to skip the chroma upsampling and YCbCr-to-RGB conversion, but since most jxl files will not actually use YCbCr, we would then also need to expose a way to know whether the output image is RGB or YCbCr.

It can all be done (we had y4m support in cjxl/djxl at some point, a long time ago before they used the API and they still had access to all the nasty codec internals), it's just a bit tricky.

As a first approximation, you could try just encoding 3 separate grayscale images, one per YCbCr component (so where the two chroma images have lower resolution). Since jxl works mostly in a planar way anyway, the total size of those 3 compressed images would be quite similar to the size of a jxl that represents the full image. If it's just to get an idea of compression performance, that would be the easiest approach (and it can also be applied to any other codec that doesn't support y4m input).
